contract.go

```diff
@@ -2,13 +2,146 @@ package exoskeleton
+}
+
+func (e CommandError) Error() string { return e.Message }
+func (e CommandError) Unwrap() error { return e.Cause }
+
+// CommandSummaryError indicates that a command did not properly implement the
+// interface for providing a summary
+type CommandSummaryError struct{ CommandError }
+
+// CommandHelpError indicates that a command did not properly implement the
+// interface for providing help
+type CommandHelpError struct{ CommandError }
+
+// CommandDescribeError indicates that an executable module did not properly
+// respond to `--describe-commands`
+type CommandDescribeError struct{ CommandError }
+
+func readSummaryFromModulefile(cmd *directoryModule) (string, error) {
+	var summary string
+
+	f, err := os.Open(cmd.path)
+	if err == nil {
+		defer f.Close()
+		summary, err = getMessageFromMagicComments(f, "summary")
+	}
+
+	if err != nil {
+		return "",
+			exit.Wrap(
+				CommandSummaryError{
+					CommandError{
+						Message: fmt.Sprintf("error reading %s: %s", cmd.path, err),
+						Command: cmd,
+						Cause:   err,
+					},
+				},
+				exit.InternalError,
+			)
+	}
+
+	return summary, nil
+}
+
+func readSummaryFromExecutable(cmd *executableCommand) (string, error) {
+	summary, err := getMessageFromCommand(cmd, "summary")
+
+	if err != nil {
+		return "",
+			exit.Wrap(
+				CommandSummaryError{
+					CommandError{
+						Message: fmt.Sprintf("error getting summary from %s: %s", cmd.path, err),
+						Command: cmd,
+						Cause:   err,
+					},
+				},
+				exit.InternalError,
+			)
+	}
+
+	return summary, nil
+}
+
+func readHelpFromExecutable(cmd *executableCommand, args []string) (string, error) {
+	var help string
+	var err error
+
+	if len(args) == 0 {
+		help, err = getMessageFromCommand(cmd, "help")
+	} else {
+		help, err = getMessageFromExecution(cmd, args, "help")
+	}
+
+	if err != nil {
+		return "",
+			exit.Wrap(
+				CommandHelpError{
+					CommandError{
+						Message: fmt.Sprintf("error getting help from %s: %s", cmd.path, err),
+						Command: cmd,
+						Cause:   err,
+					},
+				},
+				exit.InternalError,
+			)
+	}
+
+	return help, nil
+}
+
+func describeCommands(m *executableModule) (*commandDescriptor, error) {
+	cmd := m.Command("--describe-commands")
+	cmd.Stderr = nil
+	output, err := cmd.Output()
+	if err != nil {
+		return &commandDescriptor{},
+			exit.Wrap(
+				CommandDescribeError{
+					CommandError{
+						Message: fmt.Sprintf("error executing `%s --describe-commands`: %s", m.path, err),
+						Command: m,
+						Cause:   err,
+					},
+				},
+				exit.InternalError,
+			)
+	}
+
+	var descriptor *commandDescriptor
+	if err := json.Unmarshal(output, &descriptor); err != nil {
+		return &commandDescriptor{},
+			exit.Wrap(
+				CommandDescribeError{
+					CommandError{
+						Message: fmt.Sprintf("error parsing output from `%s --describe-commands`: %s", m.path, err),
+						Command: m,
+						Cause:   err,
+					},
+				},
+				exit.InternalError,
+			)
+	}
+
+	return descriptor, nil
+}
+
```

Comment:
These new wrapper methods encapsulate constructing these Command-Contract Errors.

This the code that can say, "A subcommand did not implement the contract correctly"